### ISSUE TITLE
Black image: The result of ddim_sample is black, while p_sample_loop can generate fine images.

### ISSUE THREAD

**Issue by GuHuangAI** [lucidrains/denoising-diffusion-pytorch#136]
p_sample_loop:
![generate5](https://user-images.githubusercontent.com/55237825/206614660-18e2001e-0fc4-486e-a0b2-0e8497a41c55.png)
ddim_sample:
![sample-1](https://user-images.githubusercontent.com/55237825/206614898-e59107da-ba71-446f-9e46-0b83aa7d27f4.png)
What happened?

**Comment by GuHuangAI** [comment]
I find the balck image is not caused by ddim_sample but by ema model, why does the ema model  generate wrong image? I set the ema_decay to 0.995.

**Comment by cherylLbt** [comment]
> I find the balck image is not caused by ddim_sample but by ema model, why does the ema model generate wrong image? I set the ema_decay to 0.995.

How did you save it? Did you close ema?

**Comment by cherylLbt** [comment]
> I find the balck image is not caused by ddim_sample but by ema model, why does the ema model generate wrong image? I set the ema_decay to 0.995.

How did you save it? Did you close ema? I have the same problem. Could you please share your solution? Thanks:)

**Comment by GuHuangAI** [comment]
@cherylLbt  Yeah, i closed ema. You can replace `self.ema.ema_model.sample` by `self.model.module.sample` in denoising_diffusion.py.

**Comment by cherylLbt** [comment]
> @cherylLbt Yeah, i closed ema. You can replace `self.ema.ema_model.sample` by `self.model.module.sample` in denoising_diffusion.py.

Many thanks for your reply!!🙏

**Comment by GuHuangAI** [comment]
Now, I have found the reason that, the `ema.update()` will not only update the model parameters but also change the model buffers including all the predefined buffers such as `betas`, `alphas_cumprod`, etc.
There are two solutions:
1. Do not set the predefined parameters as buffers, instead, just use `self.betas = betes` ;
2. Adding the names of predefined buffers to `param_or_buffer_names_no_ema` when defining the ema model, like this: 
`self.ema = EMA(model, param_or_buffer_names_no_ema=['betas', 'alphas_cumprod', etc])`.
I have not tried these solutions. If someone succeed, please share more info.
